File: src/main/java/dev/ddlproxy/service/DownloadService.java
package dev.ddlproxy.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;

import dev.ddlproxy.model.DownloadLinks;

import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.parser.Parser;
import org.jsoup.select.Elements;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.util.MultiValueMap;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class DownloadService {
    private static final Logger logger = LoggerFactory.getLogger(DownloadService.class);
    private final String thisBaseUrl;

    public DownloadService(String thisBaseUrl) {
        this.thisBaseUrl = thisBaseUrl;
    }

    public String handleQuery(MultiValueMap<String, String> allParams) {
        try {
            StringBuilder queryString = new StringBuilder("https://feed.animetosho.org/api?");
            List<String> qValues = new ArrayList<>();
            String seasonStr = null;
            String episodeStr = null;

            for (Map.Entry<String, List<String>> entry : allParams.entrySet()) {
                String key = entry.getKey();
                List<String> values = entry.getValue();

                if (key.equalsIgnoreCase("season") && !values.isEmpty()) {
                    seasonStr = String.format("S%02d", Integer.parseInt(values.getFirst()));
                } else if (key.equalsIgnoreCase("ep") && !values.isEmpty()) {
                    episodeStr = String.format("E%02d", Integer.parseInt(values.getFirst()));
                } else if (key.equalsIgnoreCase("q")) {
                    qValues.addAll(values);
                } else if (!key.equalsIgnoreCase("t")) {
                    for (String value : values) {
                        if (!queryString.toString().endsWith("?")) {
                            queryString.append("&");
                        }
                        queryString.append(URLEncoder.encode(key, StandardCharsets.UTF_8).replace("+", "%20"))
                                .append("=")
                                .append(URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20"));
                    }
                }
            }

            if (!qValues.isEmpty()) {
                String suffix = (seasonStr != null || episodeStr != null)
                        ? " " + (seasonStr != null ? seasonStr : "") + (episodeStr != null ? episodeStr : "")
                        : "";
                for (String value : qValues) {
                    if (!queryString.toString().endsWith("?")) {
                        queryString.append("&");
                    }
                    queryString.append("q=").append(URLEncoder.encode(value + suffix, StandardCharsets.UTF_8).replace("+", "%20"));
                }
            }

            Connection.Response response = Jsoup.connect(queryString.toString())
                    .ignoreContentType(true)
                    .parser(Parser.xmlParser())
                    .timeout(10000)
                    .execute();

            if (response.statusCode() != 200) {
                throw new IOException("Upstream returned " + response.statusCode() +
                                      " body: " + response.body());
            }

            Document doc = response.parse();

            Elements items = doc.select("item");

            for (Element item : items) {
                Element desc = item.selectFirst("description");
                if (desc == null) continue;

                Element titleElement = item.selectFirst("title");
                if (titleElement == null) continue;

                String title = titleElement.text();
                String fakeUrl = thisBaseUrl + "/download/" + URLEncoder.encode(title, StandardCharsets.UTF_8) + ".torrent";

                Elements magnetUrlElements = item.select("torznab\\:attr[name=magneturl]");
                for (Element magnetUrlElement : magnetUrlElements) {
                    magnetUrlElement.remove();
                }

                item.select("enclosure").forEach(enclosure -> enclosure.attr("url", fakeUrl));
            }

            return doc.toString();
        } catch (IOException e) {
            Document errorDoc = Document.createShell("");
            errorDoc.outputSettings().syntax(Document.OutputSettings.Syntax.xml);

            Element error = errorDoc.body().appendElement("error");
            error.appendElement("description").text(e.getMessage());

            return errorDoc.body().html();
        } catch (Exception e) {
            return "<error><description>Unexpected server error</description></error>";
        }
    }

    public DownloadLinks getLinks(String query) {
        String jsonEndpoint = "https://feed.animetosho.org/json";
        String finalUrl = jsonEndpoint + "?q=" + URLEncoder.encode(query, StandardCharsets.UTF_8);

        JsonNode searchResults = JsonNodeFactory.instance.objectNode();;

        try {
            searchResults = GET(finalUrl);
        } catch (IOException | InterruptedException e) {
            logger.error("error getting searchResults: {}", e);
        }

        if (!searchResults.isArray() || searchResults.isEmpty()) {
            logger.warn("No results found for query: {}", query);
            return new DownloadLinks(new ArrayList<>(), query);
        }

        int torrentId = searchResults.get(0).get("id").asInt();
        logger.trace("Torrent ID is: {}", torrentId);

        JsonNode torrentData = JsonNodeFactory.instance.objectNode();
        try {
            torrentData = GET(jsonEndpoint + "?show=torrent&id=" + torrentId);
        } catch (IOException | InterruptedException e) {
            logger.error("error getting data for torrent id {}: {}", torrentId, e);
        }

        JsonNode filesArray = torrentData.path("files");

        ArrayList<ArrayList<String>> links = new ArrayList<>();

        if (!filesArray.isArray()) {
            logger.error("No 'files' array found for torrent ID {}", torrentId);
            return new DownloadLinks(new ArrayList<>(), query);
        }

        for (JsonNode fileNode : filesArray) {
            JsonNode linksNode = fileNode.path("links");
            if (linksNode != null && linksNode.isObject()) {
                for (Map.Entry<String, JsonNode> entry : linksNode.properties()) {
                    JsonNode valueNode = entry.getValue();
                    ArrayList<String> innerList = new ArrayList<>();
                    if (valueNode.isArray()) {
                        for (JsonNode part : valueNode) {
                            innerList.add(part.asText());
                        }
                    } else {
                        innerList.add(valueNode.asText());
                    }
                    if (!innerList.isEmpty()) {
                        links.add(innerList);
                    }
                }
            }
        }

        logger.trace("Final links list: {}", links);

        return new DownloadLinks(links, query);
    }

    // I really tried to get RestTemplate to work here but for some reason it just didn't -_-
    private JsonNode GET(String URL) throws IOException, InterruptedException {
        logger.trace("Url: {}", URL);
        HttpResponse<String> response;
        HttpClient client = HttpClient.newBuilder().build();

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(URL))
                .GET()
                .build();

        response = client.send(request, HttpResponse.BodyHandlers.ofString());

        //logger.trace("Raw response: {}", response.body());

        ObjectMapper mapper = new ObjectMapper();
        return mapper.readTree(response.body());
    }
}

File: src/main/kotlin/dev/ddlproxy/service/DownloadService.kt
package dev.ddlproxy.service

import dev.ddlproxy.model.Release
import dev.ddlproxy.model.DownloadSource
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.util.MultiValueMap
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.*
import java.io.StringWriter
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

@Service
class DownloadService(
    private val sources: List<DownloadSource>,
    @param:Value($$"${base.url}") private val rawBaseUrl: String,
) {
    private val logger = LoggerFactory.getLogger(DownloadService::class.java)

    private val thisBaseUrl = rawBaseUrl.trim().trimEnd('/')

    suspend fun handleQuery(params: MultiValueMap<String, String>): String {
        return try {
            val qValues = mutableListOf<String>()
            var season: Int? = null
            var episode: Int? = null

            for ((key, values) in params) {
                when {
                    key.equals("season", true) && values.isNotEmpty() -> {
                        season = values.first().toIntOrNull()
                    }

                    key.equals("ep", true) && values.isNotEmpty() -> {
                        episode = values.first().toIntOrNull()
                    }

                    key.equals("q", true) -> qValues.addAll(values)
                }
            }

            if (qValues.isEmpty()) return ""

            val query = qValues.joinToString(" ")

            val results = coroutineScope {
                sources.map { source ->
                    async(Dispatchers.IO) {
                        runCatching {
                            if (query.isBlank()) {
                                source.getRecent()
                            } else {
                                source.search(query, season, episode)
                            }
                        }.getOrElse {
                            logger.warn("Source failed: {}", source.name, it)
                            emptyList()
                        }
                    }
                }.awaitAll().flatten()
            }

            buildTorznabFeed(results)

        } catch (e: Exception) {
            createErrorXml(e.message ?: "Unexpected server error")
        }
    }

    suspend fun downloadRelease(sourceName: String, identifier: String) {
        val source = sources.find { it.name.name == sourceName }

        if (source == null) {
            logger.error("Unknown source: {}", sourceName)
            return
        }

        val links = try {
            source.download(identifier)
        } catch (e: Exception) {
            logger.error("Failed to fetch links for {}:{}", sourceName, identifier, e)
            return
        }
    }

    private fun buildTorznabFeed(releases: List<Release>): String {
        val docFactory = DocumentBuilderFactory.newInstance()
        val docBuilder = docFactory.newDocumentBuilder()
        val doc = docBuilder.newDocument()

        // <rss>
        val rss = doc.createElement("rss")
        rss.setAttribute("version", "2.0")
        rss.setAttribute("xmlns:torznab", "http://torznab.com/schemas/2015/feed")
        rss.setAttribute("xmlns:atom", "http://www.w3.org/2005/Atom")
        doc.appendChild(rss)

        // <channel>
        val channel = doc.createElement("channel")
        rss.appendChild(channel)

        fun appendText(parent: org.w3c.dom.Element, name: String, value: String) {
            val el = doc.createElement(name)
            el.appendChild(doc.createTextNode(value))
            parent.appendChild(el)
        }

        appendText(channel, "title", "DDL Proxy")
        appendText(channel, "description", "Latest releases feed")

        val atomLink = doc.createElement("atom:link")
        atomLink.setAttribute("href", "$thisBaseUrl/api")
        atomLink.setAttribute("rel", "self")
        atomLink.setAttribute("type", "application/rss+xml")
        channel.appendChild(atomLink)

        appendText(channel, "link", "https://github.com/realSZ27/tosho-ddl/")
        appendText(channel, "language", "en-us")
        appendText(channel, "ttl", "30")

        releases.forEach { release ->
            val downloadUrl = buildDownloadUrl(release)

            val item = doc.createElement("item")
            channel.appendChild(item)

            appendText(item, "title", release.title)
            appendText(
                item, "pubDate",
                DateTimeFormatter.RFC_1123_DATE_TIME
                    .format(release.pubDate.atOffset(ZoneOffset.UTC))
            )

            if (release.webpageLink != null)
                appendText(item, "comments", release.webpageLink)

            fun torznabAttr(name: String, value: String) {
                val attr = doc.createElement("torznab:attr")
                attr.setAttribute("name", name)
                attr.setAttribute("value", value)
                item.appendChild(attr)
            }

            torznabAttr("category", "2020")
            torznabAttr("category", "5070")
            torznabAttr("files", "1")
            torznabAttr("seeders", "10000")
            torznabAttr("leechers", "0")
            torznabAttr("peers", "10000")

            if (release.fileSize != null)
                torznabAttr("size", release.fileSize.toString())

            val enclosure = doc.createElement("enclosure")
            enclosure.setAttribute("url", downloadUrl)
            enclosure.setAttribute("type", "application/x-bittorrent")
            enclosure.setAttribute("length", "0")
            item.appendChild(enclosure)

            appendText(item, "description", "Source: ${release.source}")
        }

        // Convert to string with XML declaration
        val transformer = TransformerFactory.newInstance().newTransformer()
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8")
        transformer.setOutputProperty(OutputKeys.METHOD, "xml")
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "no")

        val writer = StringWriter()
        transformer.transform(DOMSource(doc), StreamResult(writer))

        return writer.toString()
    }

    private fun buildDownloadUrl(release: Release): String {
        val encoded = URLEncoder.encode(
            "${release.source}|${release.identifier}",
            StandardCharsets.UTF_8
        )
        return "$thisBaseUrl/download/$encoded.torrent"
    }

    private fun createErrorXml(message: String): String {
        val docFactory = DocumentBuilderFactory.newInstance()
        val docBuilder = docFactory.newDocumentBuilder()
        val doc = docBuilder.newDocument()

        // <error>
        val error = doc.createElement("error")
        doc.appendChild(error)

        // <description>
        val description = doc.createElement("description")
        description.appendChild(doc.createTextNode(message))
        error.appendChild(description)

        // Convert to string
        val transformer = TransformerFactory.newInstance().newTransformer()
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8")
        transformer.setOutputProperty(OutputKeys.METHOD, "xml")
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "no")

        val writer = StringWriter()
        transformer.transform(DOMSource(doc), StreamResult(writer))

        return writer.toString()
    }
}